Inital State Management + Draft Applications Implementation

gittb opened the pull request:
# **Ben written:**

Adds flat file state management, implements the applications endpoint including device requests for GPUs by UUID.

While building out applications I felt more and more that /llms was a redudant endpoint and would really implement almost the same logic. I think we should remove /llms and consider a rename of /applications to have a consolidated endpoint for launching containers.

In pursuit of this, we should likely centralize where we specific the application types (and their unique params, like OpenWebUI having a volume for example) to another area rather than __init__.py as this draft implementation does.

Due to /application a couple more endpoints to get logs and such, we needed to be more verbose with the naming, I continued theme to include /create /list /{id}/delete and others, since this pattern was required to accomodate the number of endpoints in applications, I also named the endpoints in /models and /images more verbosely so they matched the pattern.

In the current state, you can grab the openwebui image via the images endpoint, and use it to create and manage a deployment via the applications endpoint. 

Finally, I ruff formatted which is why so many files are touched.

More additions below..

# **Copilot written:**

This pull request introduces significant enhancements to the InferAdmin application, focusing on asynchronous operations, logging improvements, and better thread pool management. The changes span across configuration, documentation, and core functionality.

Key changes include:

### Asynchronous Operations and Thread Pool Management:
* Added support for asynchronous operations in container management and model downloads, along with initialization and utilization of specialized thread pools for IO and CPU tasks. (`src/inferadmin/common/async_utils.py`, `src/inferadmin/common/container_management.py`, `docs/scope.md`) [[1]](diffhunk://#diff-65ab6c0813b82ff5a645c5782d10d2dccc41da60dc61d6448eb7657a37e383e5R1-R120) [[2]](diffhunk://#diff-e3733bb22135faea5dfe21cbea39409a1db4ad29ea558730877152c41418fac1R1-R173) [[3]](diffhunk://#diff-bd94d4b03da9ab69e636bc807ca892e53b532aec635152e94072322577bf5b20L6-R15)

### Logging Enhancements:
* Introduced a configurable logging system with options for log levels and output to files, replacing previous print statements. (`src/inferadmin/common/logging.py`, `src/inferadmin/lifespan.py`) [[1]](diffhunk://#diff-37a32a5a010d4c35beecd5f02399f13f7a2715ab17e33107a8d1065e77ccaa69R1-R53) [[2]](diffhunk://#diff-1192c6617606c2e7aa866cc2c0eb786feab888e3979b32296bf0680302390244R3-R44)

### Configuration Updates:
* Updated the configuration model to include new settings for thread pool sizes and logging, and modified the `.env.example` file to reflect these changes. (`src/inferadmin/config/models.py`, `src/inferadmin/config/loader.py`, `.env.example`) [[1]](diffhunk://#diff-37ed529ac09a7a936dc286d5d7aad371e84dd3766b9554e4010cfcd016dd2adaL1-R21) [[2]](diffhunk://#diff-6012e4dfbaa273509d38ea08f9aade4ae9a9b723dce967ff7f4efd3b04678526R3) [[3]](diffhunk://#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8cR3-R6)

### Documentation Updates:
* Updated the documentation to reflect the new asynchronous capabilities and thread pool configurations. (`docs/scope.md`, `docs/CNAME`) [[1]](diffhunk://#diff-bd94d4b03da9ab69e636bc807ca892e53b532aec635152e94072322577bf5b20L6-R15) [[2]](diffhunk://#diff-00d22ea612a748d8ddd4e5bf6a203591ae83569e7b16f50a11fa6d31a5beddddR1)

### Codebase Cleanup and Refactoring:
* Refactored the router and tag definitions to improve clarity and maintain backward compatibility. (`src/inferadmin/routes/__init__.py`, `src/inferadmin/openapi_tags.py`) [[1]](diffhunk://#diff-b043c189f6b940b7cf4212f439f9c1c2fe16c336d7e581b7155c5fae9223bf4fL2-R13) [[2]](diffhunk://#diff-b41dbdc5bcb9c119c8224805b9bd48e09ad92219c03a21e4e9f2068b4b037020L7-R24)

These changes collectively enhance the performance, maintainability, and user experience of the InferAdmin application.

gittb commented on src/inferadmin/routes/deployments/applications/support.py:
```diff
@@ -0,0 +1,207 @@
+        if application:
+            application.logs = logs
+            app_manager.update(application)
+        return
```
ty buddy

gittb commented:
In offline discussion decided to remove /deployments level of api, only handle container lifecycle within /applications and prepare /llms for proxying requests to the deploy inference engines

also modified /gpus endpoint to use uuid instead of index to support gpu assignment to containers via /applications

SecretiveShell commented on src/inferadmin/common/container_management.py:
```diff
@@ -0,0 +1,167 @@
+            ]
+            
+        # Launch container
+        container = DockerManager.client.containers.run(
```
maybe we can use something like ```python
loop = asyncio.get_running_loop()
loop.run_in_executor()
``` to run this in a non blocking thread

SecretiveShell commented on src/inferadmin/routes/images/__init__.py:
```diff
@@ -11,79 +11,78 @@
     """Pull a docker image and mark it as managed by InferAdmin"""
     try:
         # Use the support function to pull and tag the image
         image = pull_container_image(data.repo)
```
maybe we can use something like ```python
loop = asyncio.get_running_loop()
loop.run_in_executor()
``` to run this in a non blocking thread

gittb commented:
Added Async decorator to break out synchronous ops into different threads - Differentiated between cpu intensive ops, which really need their own thread, and IO blocked ops, which don't really need their own thread.

Also added some env args to control the thread pool size available to each space of async decorator

Added basic logger - need to be spread app wide with better scope/info management.. could be switched to loguru in the future.

Removed URL base params from endpoints in favor of payload based. 

updated scope & example .env

@SecretiveShell Should resolve above comments... please review.
